### PR TITLE
Only Construct dependencies Once when resolve is invoked

### PR DESCRIPTION
**Pull request by john-twigg-ck**
There are cases where Dip fails to check for previously resolved instances at the correct time, resulting in a dependency graph being allocated that is thrown away.

This change does the following: 
1) It migrates to use an autogenerated API so that changes to the REsolve<A,B,C> functions can be made more easily
2) These resolve calls now, prior to resolving the requested class, check's once more if the dependency has been resolved.

### CODE REVIEW

File: Sources/RuntimeArguments+Generated.swift [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+    // MARK: 3 Runtime Arguments
+
+    /// - seealso: `register(_:type:tag:factory:)`
+
+    @discardableResult
+    public func register<T, A, B, C>(_ scope: ComponentScope = .shared,
+                                     type: T.Type = T.self,
+                                     tag: DependencyTagConvertible? = nil,
+                                     factory: @escaping (A, B, C) throws -> T) -> Definition<T, (A, B, C)> {
+        var definition: Definition<T, (A, B, C)>!
```
Comment (aryan-ghassemi-ck) [inline Sources/RuntimeArguments+Generated.swift]: Did you find out if this causes a retain cycle or not

**Review by aryan-ghassemi-ck — APPROVED** [review]

File: scripts/RuntimeArguments+Generated.erb [hunk 2]
```diff
@@ -0,0 +1,110 @@
+
```
Comment (matthew-purland-ck) [inline scripts/RuntimeArguments+Generated.erb]: Fix whitespace

File: Dip/DipTests/ParentTests.swift [hunk 3]
```diff
@@ -529,8 +533,62 @@ class ParentTests: XCTestCase {
       return
     }
     XCTAssert(wireFrame.presenter.wireFrame === wireFrame)
-    XCTAssert(count == 2)
+    XCTAssert(count == 1)
     XCTAssertEqual("Title:1", wireFrame.title)
   }
 
+
```
Comment (matthew-purland-ck) [inline Dip/DipTests/ParentTests.swift]: Fix whitespace

File: Dip/DipTests/ParentTests.swift [hunk 4]
```diff
@@ -461,13 +464,14 @@ class ParentTests: XCTestCase {
     guard let levelTwoInjected = try? levelThreeContainer.resolve() as LevelTwoInjected else {
       XCTFail("Failed to create LevelTwoInjected")
       return
-    }
+  }
```
Comment (matthew-purland-ck) [inline Dip/DipTests/ParentTests.swift]: Fix indentation

File: Dip/DipTests/ParentTests.swift [hunk 5]
```diff
@@ -529,8 +533,62 @@ class ParentTests: XCTestCase {
...
     XCTAssertEqual("Title:1", wireFrame.title)
   }
 
+
+
+ /**
+ *  Esnure that once instance of the class is constructed when the class is 
+ *  registered in the child container.
+ */
+  func testOnlyConstructOnceWhenRegisteredInChildContainer(){
```
Comment (matthew-purland-ck) [inline Dip/DipTests/ParentTests.swift]: space between () and {

Comment (matthew-purland-ck) [inline Dip/DipTests/ParentTests.swift]: Fix indentation

File: scripts/RuntimeArguments+Generated.erb (hunk 2 above)
Comment (john-twigg-ck) [reply scripts/RuntimeArguments+Generated.erb]: This is consistent with the code style it was copied from.

**Review by matthew-purland-ck — APPROVED** [review]

**Comment by john-twigg-ck** [comment]
@matthew-purland-ck  has approved this PR.

**Pull request merged**